File: backend/src/main/java/tom/projects/monkeylog/common/mapper/PageMapper.java
package tom.projects.monkeylog.common.mapper;

import org.springframework.stereotype.Component;
import tom.projects.monkeylog.common.dto.Page;

import java.util.List;
import java.util.function.Function;

@Component
public class PageMapper {
    public <T, U> Page<U> map(org.springframework.data.domain.Page<T> page, Function<List<T>, List<U>> mapper) {
        return new Page<>(page.getNumber(), page.getSize(), page.getNumberOfElements(), mapper.apply(page.getContent()));
    }
}

File: backend/src/main/kotlin/tom/com/monkeylog/common/mapper/PageMapper.kt
package tom.com.monkeylog.common.mapper

import tom.com.monkeylog.common.dto.Page

fun <T, U> org.springframework.data.domain.Page<T>.toResponse(mapper: (T) -> U) =
    Page(
        isFirst = isFirst,
        isLast = isLast,
        totalPages = totalPages,
        totalElements = totalElements,
        number = number,
        size = size,
        content = content.map(mapper)
    )